src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperScopeManager.cs

```diff
@@ -124,5 +127,63 @@ public TagHelperExecutionContext End()
 
             return null;
         }
+
+        private class ExecutionContextPool
+        {
+            private readonly List<TagHelperExecutionContext> _unusedExecutionContexts;
+            private int _lastIndex;
+
+            public ExecutionContextPool()
+            {
+                _lastIndex = -1;
+                _unusedExecutionContexts = new List<TagHelperExecutionContext>();
```

Comment:
If you're using this like a stack isn't `_lastIndex` always equal to `Count - 1`